### ISSUE TITLE
faster-rcnn?

### ISSUE THREAD

**Issue by LinlyAC** [fizyr/keras-retinanet#344]
I have an easy question. 
Can I use this code to build faster-rcnn?
![1](https://user-images.githubusercontent.com/16618172/37698802-7bbcf898-2d1f-11e8-9f88-ae6f8ffc5c85.png)

**Comment by Cre1th** [comment]
@LinlyAC No you cannot, at least not end-to-end since the region proposal network generates rois which your classifier evaluates again. Since those rois are computed at runtime the generator cannot compute the target values of the regression and classification in regard to the rois beforehand. You could use alternating training and freeze parts (either rpn or classifier) each time. Since the authors themselves said that end-to-end is preferred a workaround is to move the target and loss computation into the keras graph and change the keras loss function to the identity function. This makes the model more flexible and can support more architecture changes.

I have done this and it works fine. The only downside is that when using direct methods like RetinaNet more computation (target computation) is moved from the CPU to the GPU which can make training slightly slower. But I guess this change is very minimal. Either way for Faster R-CNN it is way faster than alternating training.

I think yhenon redirected to this repo because of the performance difference but it is another architecture.

**Comment by hgaiser** [comment]
Yes, retinanet serves the same purpose as faster rcnn but it is faster and more accurate, which is why @yhenon redirected here. The architecture is different however.

@Cre1th small sidenote, you don't necessarily need to compute the loss inside the network for end2end learning. An example is here: https://github.com/fizyr/keras-maskrcnn/blob/master/keras_maskrcnn/losses.py . The targets and outputs have to be "moulded" in a certain way, which is ugly, but it is possible (and IMO preferable compared to computing inside the graph). Let's discuss further on Slack if you want.

I'll consider this issue resolved.

**Comment by Cre1th** [comment]
@hgaiser yes this is a good idea. But I think internally it is added to the keras/tensorflow graph either way, but your approach makes for a nicer reuse. I have to define IO of the submodel when I want to do predictions.
